### ISSUE TITLE
Logins are blocked when flood stage is reached

### ISSUE THREAD

**Issue by jordansissel** [elastic/elasticsearch#49049]
When the cluster flood stage is hit, all indexes appear to be set to read-only.

Problem: SAML and possibly other realms cannot login during this situation because the `.security-tokens-7` index is marked read-only.

Error message from Kibana:
```
{"statusCode":403,"error":"Forbidden","message":"[cluster_block_exception] index [.security-tokens-7] blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];"}
```

Under flood stage scenarios, attempting to use Kibana will fail. Even a super user in that realm cannot login.

It is understood that native/file realm users should be able to login, but these users may not even exist or those realms may be disabled.

A known workaround is to login as the default superuser `elastic` which does not use the affected realms. However, as this may be an uncommon task (login as `elastic`), requiring such an uncommon task in order to perform repairs is recipe, in my experience, to prevent those repairs or worse create new problems.

Ideally, non-file/native realms can login even after flood stage is reached.

**Comment by tvernum** [comment]
This specific scenario is under consideration for the propsal to formalise system indices in a future version (ideally 8.0), but I think it's a band-aid.

There are real failure scenarios, such a no elected master, or unallocated security indices, where we cannot reasonbly make SAML authentication work. We need some way to store a session token for SAML and we cannot do that in a cluster that cannot reach consensus on data.

Admins *must* have an alternative authentication mechanism, ideally file-based for those sorts of scenarios.

**Comment by jordansissel** [comment]
> Admins must have an alternative authentication mechanism

Ahh. I certainly didn't expect this. I wonder if we could call this out more strongly in our docs and with our integrations? (Example: Authy asks me for my recovery password periodically to make sure I haven't forgotten it in case of bad times; Kibana could ask super users periodically for extra safety to test their knowledge of the `elastic` password or something like that)

With no elected master, nobody's using Kibana. Similar for unallocated security indices, I imagine?

I'm thinking about a scenario where a superuser needs to make an administrative change to fix things. Are there many cases where a file-realm super user is the only accessible means to perform recovery tasks?

**Comment by tvernum** [comment]
> Are there many cases where a file-realm super user is the only accessible means to perform recovery tasks?

No, and the ES team is actively trying to make the cluster auto recover in more cases, but you can definitely get into situations where the cluster is broken and you need to authenticate in order to get diagnostic information or to enact recovery.

The primary case is just when you cluster is "broken" and no one can authenticate via Kibana (as you point out) and ingest is failing and you just want to get on and collect health/stats.
If the cluster is in such a bad state that the security index is unavailable, then a file user (with file roles, or built-in roles) may be your only option. (In more recent ES versions (5.6+) we keep the user cache alive when the cluster goes red, but that's only of use if the recovery/admin user was active in the cache).

Sometimes you need to force a reroute, or remove a block (like this issue) or some other sort of update action to get the cluster back online. In the longer term we would like clusters to be better at sorting themselves out, and not need those sorts of steps, but _right now_, if your only means to admin a cluster requires that it be yellow or better, then you have a risk.

We don't do a good job of documenting that - largely because we don't have any sort of "best practices guide", and puttin in a "setting up security" guide is too early for the reader to understand the choice they're making and the information they need to digest.

**Comment by bytebilly** [comment]
We discussed this issue and we figured out that allowing SAML logins in a flood stage will not actually be enough to allow normal user activity n the cluster. There are many other components (e.g. Kibana itself) that need to write to other indices, and there is the risk to consume the remaining space and create even worse problems.

This concern is not different with system indices, and probably the security index cannot be considered an exception. The use of a "local" administrator to fix the issue is still the best practice in this scenario.
